### DISCUSSION TITLE
HttpClient singleton and many urls

### DISCUSSION THREAD

**Discussion by okarpov** [dotnet/runtime#79220] [Q&A]
We know that Microsoft recommends using one HttpClient per application to get best performance and stabillity.

But it is not obvious if we should use single HttpClient for all different URLs or single HttpClient per each different URL.

Is anyone from developers or Microsoft can explain this?

Thank you.

**Comment by huoyaoyuan** [comment]
> But it is not obvious if we should use single HttpClient for all different URLs or single HttpClient per each different URL.

I believe it's suggested to reuse a single HttpClient as long as possible. Initilizing a HttpClient and its connection pool is quite heavy.

**Comment by gfoidl** [comment]
> use single HttpClient for all different URLs

Yep. The HttpClient uses message handlers under the covers, and these take care of the HTTP connections, etc. So just re-use the one instance.
When working together with DI, you can also use "typed clients" if possible.

**Comment by okarpov** [comment]
yes, you all are right according to the specification perspective.

however, our observation using real world usecases seems not so obvious.

first - it looks like linux and windows behave differently.
second - seems using HttpClient per URL (which are completely different like server.com, anotherserver.org, etc...) leads to less HTTPS errors like decryption error, closed connection, timeout, 0 bytes data, etc.


that is why we rised such question to get deep and confident answers/explanaitions

**Comment by MihaZupan** [comment]
> But it is not obvious if we should use single HttpClient for all different URLs or single HttpClient per each different URL.

Use a single instance unless you actually need multiple (e.g. you need different HttpClient settings per host).
If you create an instance per host, that's also fine, you're just complicating your code and wasting some memory for no upside.

> it looks like linux and windows behave differently

How so? There shouldn't be any differences between platforms when comparing a shared HttpClient instance to one per host.

> seems using HttpClient per URL (which are completely different like server.com, anotherserver.org, etc...) leads to less HTTPS errors like decryption error, closed connection, timeout, 0 bytes data, etc.

It shouldn't. How did you determine this?

**Reply by okarpov** [reply]
sorry for the late reply, one main difference is that it uses OpenSSL in Linux
also we see more errors in Linux with HTTPS then in Windows 11

**Reply by okarpov** [reply]
and multithreading is handled differently by OpenSSL and Windows based realisation of SSL.

Also, we did measure QPS using single instance of HttpClient and separate instance of HttpClient per host and we see different results on Windows and on Linux
